### DISCUSSION TITLE
What does `outer` and `inner` do?

### DISCUSSION THREAD

**Discussion by Eisverygoodletter** [pest-parser/pest#918] [Q&A] [answered]
I have seen some code in the examples like 
```rust
#[pest_ast(outer(with(span_into_str), with(str::parse), with(Result::unwrap)))]
```
with multiple items like `outer`, `inner` and `with`. I get that outer and inner are for extracting information from a `Span`, but what are their parameters meant to be?
Also, are there any more of them I'm unaware of?

**Comment by tomtau** [comment, marked as answer]
Their parameters define the conversion strategy, so in that example of csv: https://github.com/pest-parser/ast/blob/09255d748b63ecd5741cefb5275763323871f898/derive/README.md?plain=1#L280

it'll first call `span_into_str`, then `parse`, and finally `unwrap` to get that `f64` value
